### DISCUSSION TITLE
How big is A5_VOICE segment?

### DISCUSSION THREAD

**Discussion by mungewell** [patois/A5Pack#6] [Ideas]
I was apparently using a truncated file in #1 so I set off to find the **whole** segment.... what an idiot ;-)

code says
```
    VALID_CHUNKS = { # Flash dst addr, max valid size
        "A5 code":(0x40020000, 0x200000),
        "A5 voice":(0x40200000, 0x5000000),
        "A5 Reserve1":(0x479C0000, 0x100000),
```

In going for overkill I did the following on my Explorer
![Screenshot 2024-06-03 113726](https://github.com/patois/A5Pack/assets/236907/7a41dac9-18e2-4603-8128-402897646fb7)

This 999999999 byte file seems to contain multiple identical copies. I assume that the upper address pins are not decoded and the Flash memory reports multiple times.
```
$ dd count=1024 bs=1 if=a5_voice_999999999.bin of=chunk.bin
1024+0 records in
1024+0 records out
1024 bytes (1.0 kB, 1.0 KiB) copied, 0.0221006 s, 46.3 kB/s

$ python3 ~/SearchBin-github/searchbin.py -f chunk.bin a5_voice_999999999.bin 
Match at offset:              0            0 in  a5_voice_999999999.bin
Match at offset:       67108864      4000000 in  a5_voice_999999999.bin
Match at offset:      134217728      8000000 in  a5_voice_999999999.bin
Match at offset:      201326592      C000000 in  a5_voice_999999999.bin
Match at offset:      268435456     10000000 in  a5_voice_999999999.bin
Match at offset:      335544320     14000000 in  a5_voice_999999999.bin
Match at offset:      402653184     18000000 in  a5_voice_999999999.bin
Match at offset:      469762048     1C000000 in  a5_voice_999999999.bin
Match at offset:      536870912     20000000 in  a5_voice_999999999.bin
Match at offset:      603979776     24000000 in  a5_voice_999999999.bin
Match at offset:      671088640     28000000 in  a5_voice_999999999.bin
Match at offset:      738197504     2C000000 in  a5_voice_999999999.bin
Match at offset:      805306368     30000000 in  a5_voice_999999999.bin
Match at offset:      872415232     34000000 in  a5_voice_999999999.bin
Match at offset:      939524096     38000000 in  a5_voice_999999999.bin
```

So is the true size 67108864 (0x4000000) bytes?

**Comment by mungewell** [comment]
From Explorer
```
$ ls -la a5_voice_0x4000000.bin 
-rw-rw-r-- 1 simon simon 67108864 Jun  3 12:50 a5_voice_0x4000000.bin

$ md5sum a5_voice_0x4000000.bin
8041a6599884e31f800c879f033c8fe7  a5_voice_0x4000000.bin
```

**Comment by mungewell** [comment]
There's a lot(!!) of data to view, luckily we can use the fact that `hexdump` abbreviates repeated lines to location the start of 'new' sections
```
$ grep -e "^*" -B 1 -A 5 hexdump.txt | vim -
``` 

### Observation 1
This data contains the patches from offset 0x02000000, this also means that the MD5Sum of the block is useless.
```
*
02000000  6a 00 00 00 0d 53 61 77  70 72 65 73 73 69 76 65  |j....Sawpressive|
02000010  20 47 44 00 20 00 20 00  b0 04 00 00 04 00 20 00  | GD. . ....... .|
02000020  00 00 00 00 00 00 00 00  02 00 01 00 73 00 00 00  |............s...|
02000030  00 00 00 00 00 00 00 00  01 00 01 00 00 00 00 00  |................|
02000040  00 00 00 00 00 00 00 00  00 00 00 00 00 00 03 00  |................|
```
There are 5 banks of 128 patches... and can be (minimally) decoded with
[a5_voice.py.txt](https://github.com/user-attachments/files/15540938/a5_voice.py.txt)
```
$ python3 a5_voice.py
Container: 
    presets = ListContainer: 
        ListContainer: 
            Container: 
                bits = b'j\x00' (total 2)
                number = 0
                un1 = 13
                name = 'Sawpressive GD' (total 14)
                un2 = b'\x00 \x00 \x00\xb0\x04\x00\x00\x04\x00 \x00\x00\x00\x00'... (truncated, total 2767)
            Container: 
...
            Container: 
                bits = b'\x963' (total 2)
                number = 639
                un1 = 0
                name = 'Patches by RA' (total 13)
                un2 = b'  \x00 \x00\xb0\x04\x00\x00\x04\x00 \x00\t\x00\x00'... (truncated, total 2767)
block length 1783040 bytes
```

**Comment by mungewell** [comment]
```
000674e0  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
00067500  00 00 17 00 30 00 48 00  5e 00 76 00 8e 00 a6 00  |....0.H.^.v.....|
00067510  bc 00 d5 00 ed 00 05 01  1d 01 35 01 4c 01 63 01  |..........5.L.c.|
00067520  7b 01 93 01 aa 01 c2 01  da 01 f2 01 09 02 20 02  |{............. .|
00067530  38 02 50 02 67 02 7f 02  97 02 af 02 c6 02 df 02  |8.P.g...........|
00067540  f8 02 0d 03 23 03 3c 03  55 03 6c 03 84 03 9d 03  |....#.<.U.l.....|
--
0006b500  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0006b520  00 00 30 00 5f 00 8e 00  bd 00 ed 00 1d 01 4c 01  |..0._.........L.|
0006b530  7b 01 ab 01 da 01 09 02  38 02 67 02 97 02 c7 02  |{.......8.g.....|
0006b540  f7 02 24 03 54 03 84 03  b4 03 e3 03 12 04 41 04  |..$.T.........A.|
0006b550  71 04 a0 04 cf 04 fe 04  2d 05 5d 05 8c 05 bb 05  |q.......-.].....|
0006b560  ea 05 1a 06 49 06 78 06  a7 06 d6 06 05 07 34 07  |....I.x.......4.|
--
0006d520  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
0006d540  00 00 5f 00 bd 00 1d 01  7b 01 da 01 38 02 97 02  |.._.....{...8...|
0006d550  f6 02 54 03 b4 03 12 04  71 04 cf 04 2d 05 8c 05  |..T.....q...-...|
0006d560  ea 05 49 06 a7 06 05 07  63 07 c1 07 1f 08 7d 08  |..I.....c.....}.|
0006d570  da 08 39 09 96 09 f3 09  51 0a ae 0a 0c 0b 69 0b  |..9.....Q.....i.|
0006d580  c5 0b 22 0c 7f 0c dc 0c  39 0d 95 0d f1 0d 4d 0e  |..".....9.....M.|
```

```
$ dd skip=423168 count=8192 bs=1 if=a5_voice_0x4000000.bin of=test.raw
8192+0 records in
8192+0 records out
8192 bytes (8.2 kB, 8.0 KiB) copied, 0.0745105 s, 110 kB/s
$  sox -r 48k -e signed -b 32 -c 1 --endian little test.raw test.wav
```
![Screenshot_2024-06-03_18-53-06](https://github.com/patois/A5Pack/assets/236907/27f1fd20-9ea9-4d40-bd93-3b8d34f9c896)

```
$ dd skip=439584 count=8192 bs=1 if=a5_voice_0x4000000.bin of=test.raw
8192+0 records in
8192+0 records out
8192 bytes (8.2 kB, 8.0 KiB) copied, 0.0700892 s, 117 kB/s
$  sox -r 48k -e signed -b 32 -c 1 --endian little test.raw test.wav
```
![Screenshot_2024-06-03_18-53-51](https://github.com/patois/A5Pack/assets/236907/f5af7a9e-7be0-4288-aab9-33b959581056)

```
$ dd skip=447808 count=8192 bs=1 if=a5_voice_0x4000000.bin of=test.raw
8192+0 records in
8192+0 records out
8192 bytes (8.2 kB, 8.0 KiB) copied, 0.0428686 s, 191 kB/s
$  sox -r 48k -e signed -b 32 -c 1 --endian little test.raw test.wav
```
![Screenshot_2024-06-03_18-54-54](https://github.com/patois/A5Pack/assets/236907/3117e958-426e-4ae6-a812-387e9cac6730)

Not sure whether, or how, these are mapped to the Wave-Tables, but it's very interesting...

**Comment by mungewell** [comment]
It seems there are different sized 'waveforms' within the segment, they can be 16348, 8192 or 4096 (maybe others) and are each followed by 2 rows of Zeros (32 bytes).

These are somewhat scatter through the segment, I have just been cherry picking at the moment. I suspect that there is probably a table of where these 'waveforms' are located....
```
-rw-rw-r-- 1 simon simon  4096 Jun  3 21:10 0x00d5e940.raw
-rw-rw-r-- 1 simon simon  4096 Jun  3 21:10 0x00d5f960.raw
-rw-rw-r-- 1 simon simon  4096 Jun  3 21:10 0x00d60980.raw
-rw-rw-r-- 1 simon simon  4096 Jun  3 21:10 0x00d619a0.raw
-rw-rw-r-- 1 simon simon  4096 Jun  3 21:10 0x00d629c0.raw
-rw-rw-r-- 1 simon simon  4096 Jun  3 21:10 0x00d639e0.raw
-rw-rw-r-- 1 simon simon  4096 Jun  3 21:10 0x00d64a00.raw

$ python3 ~/SearchBin-github/searchbin.py -p "004ad6" a5_voice_0x4000000.bin 
Match at offset:         256100        3E864 in  a5_voice_0x4000000.bin
Match at offset:         256104        3E868 in  a5_voice_0x4000000.bin

$ python3 ~/SearchBin-github/searchbin.py -p "e039d6" a5_voice_0x4000000.bin 
Match at offset:         256000        3E800 in  a5_voice_0x4000000.bin
Match at offset:         256004        3E804 in  a5_voice_0x4000000.bin
Match at offset:       16323232       F912A0 in  a5_voice_0x4000000.bin
Match at offset:       16331424       F932A0 in  a5_voice_0x4000000.bin
Match at offset:       16339648       F952C0 in  a5_voice_0x4000000.bin
Match at offset:       16347840       F972C0 in  a5_voice_0x4000000.bin

$ python3 ~/SearchBin-github/searchbin.py -p "c029d6" a5_voice_0x4000000.bin 
Match at offset:         255900        3E79C in  a5_voice_0x4000000.bin
Match at offset:         255904        3E7A0 in  a5_voice_0x4000000.bin

$ python3 ~/SearchBin-github/searchbin.py -p "8009d6" a5_voice_0x4000000.bin 
Match at offset:         255700        3E6D4 in  a5_voice_0x4000000.bin
Match at offset:         255704        3E6D8 in  a5_voice_0x4000000.bin
```

**Comment by mungewell** [comment]
I managed to do a crude parsing of this table, and extract ALL the waves from the segment...
[a5_voice.py.txt](https://github.com/user-attachments/files/15555898/a5_voice.py.txt)
```
$ python3 a5_voice.py
Found wave @ 0x000634e0
Found wave @ 0x00067500
Found wave @ 0x0006b520
Found wave @ 0x0006d540
Found wave @ 0x0006e560
Found wave @ 0x0006f580
Found wave @ 0x000705a0
Found wave @ 0x000715c0
Found wave @ 0x000725e0
Found wave @ 0x00073600
...

$ mkdir wav
$ find . -name "*.raw" -exec bash -c "sox -r 48k -e signed -b 32 -c 1 --endian little {} wav/{}.wav" \;
$ find wav/ -name "*_0.raw.wav" -exec bash -c "sndfile-waveform --no-rms -B 0xff000000  {} {}.wav.png" \;
```

There are 16 wavefiles for each selectable waveform, these are of different sizes (16K, 8K and 4K) and have 2 cycles - presumably so it's easy to adjust the phase.
```
-rw-rw-r--  1 simon simon   16464 Jun  4 08:07 0x01508640_219_0.raw.wav
-rw-rw-r--  1 simon simon   16464 Jun  4 08:07 0x0150c660_219_1.raw.wav
-rw-rw-r--  1 simon simon    8272 Jun  4 08:07 0x01510680_219_2.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x015126a0_219_3.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x015136c0_219_4.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x015146e0_219_5.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x01515700_219_6.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x01516720_219_7.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x01517740_219_8.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x01518760_219_9.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x01519780_219_10.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x0151a7a0_219_11.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x0151b7c0_219_12.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x0151c7e0_219_13.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x0151d800_219_14.raw.wav
-rw-rw-r--  1 simon simon    4176 Jun  4 08:07 0x0151e820_219_15.raw.wav
```

Here's a zip of images of the primary ('*_0.raw') waves.
[images.zip](https://github.com/user-attachments/files/15555393/images.zip)


Really the table should be better understood, and parsed correctly.

**Comment by mungewell** [comment]
Why 16 wave files per wave form? I looks like there is some filtering done at different amounts, not sure why they need that....
![montage](https://github.com/patois/A5Pack/assets/236907/5d56563e-afd3-4105-9432-d90eed167877)

**Reply by bozzy-gh** [reply]
> Why 16 wave files per wave form? I looks like there is some filtering done at different amounts, not sure why they need that.... !

To avoid aliasing as you move into the higher registers?
BTW, the Hydrasynth is a monster, but they didn't manage to avoid aliasing when using mutants... Terrible!

**Comment by mungewell** [comment]
A more robust decode of the table for the waveforms. Using `Const()` so we can high-light which data is actually changing...
[a5_voice.py.txt](https://github.com/user-attachments/files/15572159/a5_voice.py.txt)

**Comment by mungewell** [comment]
Looping back to the original Q, how big is the 'A5_Voice' segment....?

I noticed that there's code in the copy I made, and if I take the first 1K of the FW image it is located in my copy. Surely that doesn't count as 'Voice'.

```
$ dd -if=Explorer_v2.0.0/3M200_00000087_A5_code -of=chunk.bin -bs=1024 -count=1
$ python3 ~/SearchBin-github/searchbin.py -f chunk.bin a5_voice_0x4000000.bin
Match at offset:       65142784      3E20000 in  a5_voice_0x4000000.bin
```

The last patch appears to start here:
```
*
021b2a10  00 00 00 00 00 00 00 00  00 00 00 00 00 00 96 33  |...............3|
021b2a20  7f 02 00 50 61 74 63 68  65 73 20 62 79 20 52 41  |...Patches by RA|
```
and end here:
```
02400a20  22 32 00 00 0c 00 00 82  10 00 00 82 88 00 00 80  |"2..............|
02400a30  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
02408000  00 00 00 eb 44 00 00 eb  2c 00 8f e2 00 0c 90 e8  |....D...,.......|
```

Noting that there's a bit of space up-to `0x02408000`.

Also Note: Banks F, G & H are 'init' clear on my machine....

Maybe used in Delux (space for another 800 patches?), or useful for additional wave data if we'd want to expand the table for user waves....

**Comment by mungewell** [comment]
Another observation, data seems to start with a pointer to the first raw sample (_0_0.raw 16k)
```
00000000  c2 02 00 00 70 56 34 12  e0 34 06 00 dc 00 00 00  |....pV4..4......|
                                               ??               count of 220 wavform sets?
                                   ^^ ^^ ^^ ^^                  pointer to first waveform
                                                                0x000634e0_0_0.raw
00000010  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
00000020  89 00 00 00 42 02 00 00  43 02 00 00 44 02 00 00  |....B...C...D...|
```

And this is repeated after the raw data...
```
01501550  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
                                                                end of 0x01500540_218_9.raw
*
                                                                last in table 0x0151f840_219_16.raw
015ffff0  00 00 00 00 00 00 00 00  00 00 00 00 e0 34 06 00  |.............4..|
                                               ^^ ^^ ^^ ^^      pointer to first waveform
                                                                0x000634e0_0_0.raw
```

So maybe in truth the waveform data can't exceed this location.... :-(

**Comment by mungewell** [comment]
Improved script with `--unpack` and `--pack` functions for raw samples
[a5_voice.py.txt](https://github.com/user-attachments/files/15688403/a5_voice.py.txt)

Although the samples are noted to be 16384, 8192 or 4096 bytes, there are sometimes a couple of bytes after them. These possibly could be a checksum of some form....
```
01501530  69 0e 2a 14 7c 19 40 1e  5c 22 ba 25 47 28 f5 29  |i.*.|.@.\".%G(.)|
01501540  ba 2a 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |.*..............|
          ^^ ^^
01501550  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
```

**Reply by mungewell** [reply]
These two bytes do NOT change with the content of the sample, they seem to be similar value withing the group (ie representing the same waveform).
[checksum_with_md5.txt](https://github.com/user-attachments/files/15741388/checksum_with_md5.txt)

**Comment by mungewell** [comment]
I have made another observation/realization; the size of the 'A5_Voice' segment may be limited on the Explorer (that I have)...

For the 'Keyboard'
```
        "A5 code":          (0x40020000, 0x00200000),
        "A5 voice":         (0x40200000, 0x05000000),
        "A5 patch":         (0x47F00000, 0x000A5400),
```

But the 'Explorer has
```
        "A5_3m patch":      (0x42200000, 0x00200000),
```

Which lines up with hexdump that I have observed.
```
015ffff0  00 00 00 00 00 00 00 00  00 00 00 00 e0 34 06 00  |.............4..|
                                               ^^ ^^ ^^ ^^      pointer to first waveform
                                                                0x000634e0_0_0.raw
01600000  00 00 00 00 00 00 00 00  00 00 00 00 00 00 00 00  |................|
*
02000000  6a 00 00 00 0d 53 61 77  70 72 65 73 73 69 76 65  |j....Sawpressive|
02000010  20 47 44 00 20 00 20 00  b0 04 00 00 04 00 20 00  | GD. . ....... .|
```

The 'footer' (that last pointer) may well be in a different place for the more expensive keyboard/desktop/delux. I suspect that the memory was 'squeezed' for the Explorer so that it could fit into a single (or fewer) IC's on the board - as a cost saving measure.

This might mean that the others have more space available for waveforms, or even that they have a different structure for the samples.... **I'd love for someone to pull the A5_Voice data from one of the others and compare to what I have seen.**

**Reply by patois** [reply]
I've found an old dump of the Desktop's A5 Voice segment made in March '23. Its size is 83.886.080 bytes.
Disassembling / decompiling HydrasynthUpdaterForDevelopment.exe for every HS type (Desktop, Keyboard, ...) and finding the string "A5 Voice", then navigating to its cross references will land you in a structure that has a pointer to the segment name ("A5 Voice") followed by a 32bit uint (which I think is an address if I recall) and another 32bit uint which is the size of the segment. Hope this helps!

**Reply by mungewell** [reply]
Thanks for the reply - but I have not be dis-assembling code (not my skill set), instead I figured out what I know from looking at the **contents** of the A5Voice segment itself. My parser is based 100% on those observations, and 'walks' the data blob to extract the audio waveforms.

84MByte seems very large in comparison (x4) to the block I see from the Explorer, so perhaps they are totally different...

**Reply by patois** [reply]
I probably must have extracted the segment's size from the updater. it matches this https://github.com/patois/A5Pack/blob/13ddfb3dacd5e28ccaeac081f26763a89113183b/A5Pack.py#L20C33-L20C42

Also, I just double checked and confirmed the size of the A5 Voice segment is 83.886.080 bytes with the HS Explorer version too, according to its HydrasynthUpdaterForDevelopment.exe. This could still be a maximum total supported size whereas in reality, the actual segment/data on the device could be smaller but I didn't really reverse engineer any of that code either.

**Reply by mungewell** [reply]
I decide to move my script to it's own project, where I think it would be more suited. Thank you for your original information and help.
https://github.com/mungewell/A5Voice

**Reply by patois** [reply]
Absolutely, it's your own work and it fully deserves its own home! I'll be following your progress, keep up the great work!
